src/H5FDint.c

```diff
@@ -260,148 +300,2075 @@ H5FD_write(H5FD_t *file, H5FD_mem_t type, haddr_t addr, size_t size, const void
+
+        HDassert(file->base_addr > 0);
+
+        for (i = 0; i < count; i++) {
+
+            offsets[i] -= file->base_addr;
+        }
+    }
+
+    /* Cleanup dataspace arrays */
+    if (mem_spaces != mem_spaces_static)
+        mem_spaces = H5MM_xfree(mem_spaces);
+    if (file_spaces != file_spaces_static)
+        file_spaces = H5MM_xfree(file_spaces);
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5FD_read_selection_id() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5FD__write_selection_translate
+ *
+ * Purpose:     Translates a selection write call to a vector write call
+ *              if vector writes are supported, or a series of scalar
+ *              write calls otherwise.
+ *
+ * Return:      Success:    SUCCEED
+ *                          All writes have completed successfully.
+ *
+ *              Failure:    FAIL
+ *                          One or more writes failed.
+ *
+ * Programmer:  NAF -- 5/13/21
+ *
+ * Changes:     None
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5FD__write_selection_translate(H5FD_t *file, H5FD_mem_t type, hid_t dxpl_id, uint32_t count,
+                                H5S_t **mem_spaces, H5S_t **file_spaces, haddr_t offsets[],
+                                size_t element_sizes[], const void *bufs[])
+{
+    hbool_t         extend_sizes = FALSE;
+    hbool_t         extend_bufs  = FALSE;
+    uint32_t        i;
+    size_t          element_size;
+    const void *    buf;
+    hbool_t         use_vector = FALSE;
+    haddr_t         addrs_static[8];
+    haddr_t *       addrs = addrs_static;
+    size_t          sizes_static[8];
+    size_t *        sizes = sizes_static;
+    const void *    vec_bufs_static[8];
+    const void **   vec_bufs = vec_bufs_static;
+    hsize_t         file_off[H5FD_SEQ_LIST_LEN];
+    size_t          file_len[H5FD_SEQ_LIST_LEN];
+    hsize_t         mem_off[H5FD_SEQ_LIST_LEN];
+    size_t          mem_len[H5FD_SEQ_LIST_LEN];
+    size_t          file_seq_i;
+    size_t          mem_seq_i;
+    size_t          file_nseq;
+    size_t          mem_nseq;
+    size_t          io_len;
+    size_t          nelmts;
+    hssize_t        hss_nelmts;
+    size_t          seq_nelem;
+    H5S_sel_iter_t *file_iter      = NULL;
+    H5S_sel_iter_t *mem_iter       = NULL;
+    hbool_t         file_iter_init = FALSE;
+    hbool_t         mem_iter_init  = FALSE;
+    H5FD_mem_t      types[2]       = {type, H5FD_MEM_NOLIST};
+    size_t          vec_arr_nalloc = sizeof(addrs_static) / sizeof(addrs_static[0]);
+    size_t          vec_arr_nused  = 0;
+    herr_t          ret_value      = SUCCEED;
+
+    FUNC_ENTER_NOAPI(FAIL)
+
+    /* Sanity checks */
+    HDassert(file);
+    HDassert(file->cls);
+    HDassert(vec_arr_nalloc == sizeof(sizes_static) / sizeof(sizes_static[0]));
+    HDassert(vec_arr_nalloc == sizeof(vec_bufs_static) / sizeof(vec_bufs_static[0]));
```

Comment:
Why not give a name to the constant, `8`, like the library has named a constant H5FD_SEQ_LIST_LEN? Then the sizes of `sizes_static`, `vec_bufs_static`, and `addrs_static` can be tied together by the constant in the same way as `file_off`, `file_len`, etc. This assertion rigamarole can go away.